### PR TITLE
Range support in window expressions (support them as groups)

### PR DESCRIPTION
**Pull request by kgyrtkirk**
### Description

* support `groups`  windowing mode; which is a close relative of `ranges` (but not in the standard)
* all windows with range expressions will be executed wit it `groups`
* it will be 100% correct in case for both bounds its true that: `isCurrentRow() || isUnBounded()` 
   * this covers `OVER ( ORDER BY COL )`
* for other cases it will have some chances of getting correct results...

  

#### Release note

This PR has:

- [ ] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java [hunk 1]
```diff
@@ -85,10 +88,234 @@ public RowsAndColumns aggregateAll(
...
+        {
+          return current < b;
+        }
+      };
+
+    }
+  }
+
+  /**
+   * Represents a range between U (inclusive) and V (exclusive).
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java:239]: What is `U` and `V`?  I see `outputRows` and `inputRows`?

File: processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java [hunk 2]
```diff
@@ -85,10 +88,234 @@ public RowsAndColumns aggregateAll(
...
+  {
+    private AggregatorFactory[] aggFactories;
+    Interval currentRows = new Interval(0, 0);
+    private final AtomicInteger rowIdProvider;
+    private final ColumnSelectorFactory columnSelectorFactory;
+
+    private final Aggregator[] aggregators;
+
+
+    AggCell(AppendableRowsAndColumns rac, AggregatorFactory[] aggFactories)
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java:263]: Instead of this depending on an `AppendableRowsAndColumns` I'd suggest that you make it depend on a `ColumnSelectorFactoryMaker`.  On top of that, if it was me, I'd write it as a static creator method and have this constructor take the `AtomicInteger, ColumnSelectorFactory, AggregatorFactory[]` tuple.

File: processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java [hunk 3]
```diff
@@ -85,10 +88,234 @@ public RowsAndColumns aggregateAll(
...
+    public void moveTo(Interval newRows)
+    {
+      // incremental addition of additional values
+      if (currentRows.a == newRows.a && currentRows.b < newRows.b) {
+        for (int i = currentRows.b; i < newRows.b; i++) {
+          aggregate(i);
+        }
+        currentRows = newRows;
+        return;
+      }
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java:293]: In reading this method, I ended up getting stuck a bit because I missed this `return` statement.  Given that this method is either one or the other, an `if/else` would've likely saved me a bit of time in realizing that it was an either-or branch.

File: processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java [hunk 4]
```diff
@@ -85,10 +88,234 @@ public RowsAndColumns aggregateAll(
...
+      currentRows = newRows;
+    }
+
+    public void setOutputs(Object[][] results, Interval outputRows)
+    {
+      for (int aggIdx = 0; aggIdx < aggFactories.length; aggIdx++) {
+        Object aggValue = aggregators[aggIdx].get();
+        for (int rowIdx = outputRows.a; rowIdx < outputRows.b; rowIdx++) {
+          results[aggIdx][rowIdx] = aggValue;
+        }
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java:307]: If, e.g., this method is called, then `aggregate()` is called before `newAggregators()` is called, it's totally acceptable (though not currently documented on the `Aggregator` interface) that the underlying object itself gets mutated as part of the subsequent `aggregate()` call.  For a cumulative aggregation, you need to do a `.get()` and then feed that as the first input back into a new Aggregator object.

File: processing/src/test/java/org/apache/druid/query/rowsandcols/semantic/FramedOnHeapAggregatableTest.java [hunk 5]
```diff
@@ -456,4 +459,140 @@ public void testReverseCumulativeAggregation()
...
+        }
+    );
+
+    new RowsAndColumnsHelper()
+    .expectColumn("c1", c1Vals)
+    .expectColumn("c2", c2Vals)
+    .expectColumn("res", resVals)
+        .allColumnsRegistered()
+        .validate(results);
+  }
```
Comment (imply-cheddar) [inline processing/src/test/java/org/apache/druid/query/rowsandcols/semantic/FramedOnHeapAggregatableTest.java:596]: It might be good to also have some tests using the HLL aggregators.  I expect them to be a good thing that exercises the whole "the object gets mutated after `.get()`" thing that I mentioned in a different comment.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteWindowQueryTest.java [hunk 6]
```diff
@@ -64,7 +64,7 @@ public class CalciteWindowQueryTest extends BaseCalciteQueryTest
 
   public static final boolean DUMP_ACTUAL_RESULTS = Boolean.parseBoolean(
       System.getProperty("druid.tests.sql.dumpActualResults")
-  );
+  ) || true;
```
Comment (imply-cheddar) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteWindowQueryTest.java:67]: I'm thinking that you added this just for simplicity while developing and meant to remove this before merge?

File: sql/src/test/resources/calcite/tests/window/range_handling.sqlTest [hunk 7]
```diff
@@ -0,0 +1,18 @@
+type: "operatorValidation"
+
+sql: |
+  SELECT
+    FLOOR(m1/3),
+    --DENSE_RANK() OVER (ORDER BY FLOOR(m1/3)),
```
Comment (imply-cheddar) [inline sql/src/test/resources/calcite/tests/window/range_handling.sqlTest:6]: Remove the comments before merge?

**Review by imply-cheddar — COMMENTED** [review]
I did an initial pass over it and left various comments here and there.  Looking back at it, the comments aren't really focused on the meat of the logic but on the structuring.  I think I likely need to spend more time internalizing what the whole `inputRows` and `outputRows` things are on the `Range` object to really believe I understand the meat.

File: processing/src/main/java/org/apache/druid/query/operator/window/WindowFrame.java [hunk 8]
```diff
@@ -44,21 +49,35 @@ public enum PeerType
...
+
+  public WindowFrame(
+      PeerType peerType,
+      boolean lowerUnbounded,
+      int lowerOffset,
+      boolean upperUnbounded,
+      int upperOffset)
+  {
+    this(peerType, lowerUnbounded, lowerOffset, upperUnbounded, upperOffset, null);
+  }
```
[the remark this answers is not in the document]
Comment (kgyrtkirk) [reply processing/src/main/java/org/apache/druid/query/operator/window/WindowFrame.java:63]: sure - I can reduce the number of constructors; but in this case I would also remove the extra boolean for the presence of lower/upper; as it just bloats the json representation....possible alternatives could be to use:
* custom reader/writer
* or...use integer and treat `null` as unbounded

File: processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java [hunk 9]
```diff
@@ -85,10 +88,234 @@ public RowsAndColumns aggregateAll(
...
+
+    public RangeIteratorForWindow(AppendableRowsAndColumns rac, WindowFrame frame)
+    {
+      assert (frame.getPeerType() == PeerType.RANGE);
+      rangeToRowId = ClusteredGroupPartitioner.fromRAC(rac).computeBoundaries(frame.getOrderByColNames());
+      numRows = rac.numRows();
+      numRanges = rangeToRowId.length - 1;
+      lowerOffset = frame.getLowerOffsetClamped(numRanges);
+      upperOffset = frame.getUpperOffsetClamped(numRanges) + 1;
+    }
```
[the remark this answers is not in the document]
Comment (kgyrtkirk) [reply processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java:133]: I think its a good think to make the contract here that this class will provide an iterator based on the Frame for that given RAC

passing `int[]` would make the contract more complicated

I think this same call pattern can be reused to provide other iterators as well - so that it becomes the iterators job to select the right rowranges

File: processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java [hunk 10]
```diff
@@ -85,10 +88,234 @@ public RowsAndColumns aggregateAll(
...
+    private final int numRanges;
+    private final int lowerOffset;
+    private final int upperOffset;
+
+    public RangeIteratorForWindow(AppendableRowsAndColumns rac, WindowFrame frame)
+    {
+      assert (frame.getPeerType() == PeerType.RANGE);
+      rangeToRowId = ClusteredGroupPartitioner.fromRAC(rac).computeBoundaries(frame.getOrderByColNames());
+      numRows = rac.numRows();
+      numRanges = rangeToRowId.length - 1;
```
[the remark this answers is not in the document]
Comment (kgyrtkirk) [reply processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java:130]: this is number of ranges `rangeToRowId` is:
```
[ range1_start, range1_start, ...., rangeN_start, numRows ].length == N + 1
```

File: processing/src/main/java/org/apache/druid/query/operator/window/WindowFramedAggregateProcessor.java [hunk 11]
```diff
@@ -96,4 +96,32 @@ public String toString()
...
+    }
+    if (obj == null) {
+      return false;
+    }
+    if (getClass() != obj.getClass()) {
+      return false;
+    }
+    WindowFramedAggregateProcessor other = (WindowFramedAggregateProcessor) obj;
+    return Arrays.equals(aggregations, other.aggregations) && Objects.equals(frame, other.frame);
+  }
```
[the remark this answers is not in the document]
Comment (kgyrtkirk) [reply processing/src/main/java/org/apache/druid/query/operator/window/WindowFramedAggregateProcessor.java:124]: There are tests which use these - its a bit hard to take a look at a list of operators without clear context about what the named fields will contain...without context that makes it pretty hard to follow - and may even need a complicated detour [like in this PR](https://github.com/apache/druid/pull/14007#discussion_r1363362711)

I understand the intention but I don't think these equiv things will be that helpfull in the longrun:

I wonder if
`select col as u,1 as v from t`
should be considered equivalent to
`select col as v,1 as u from t`
?
probably...

what if its being used as a subquery:
`select substring(u,v) from T`
then I guess they are clearly not equiv

w.r.t. to window processors; is it the same:
```
rank() as u,
row_number() as v
```
and
```
rank() as v,
row_number() as u
```
?
I guess right now that would return `true` - but I believe that they equivalence depends on the outer context not just on the expression.

To get real equivalence level comparision the whole tree should be either rewritten or compared via some stuff (observer?) which could register and could keep track equiv pieces; while retaining the concept of left and right - but that's now how these `validateEquivalent` calls are designed

I don't know what benefits are we getting from taking the above risks `validateEquivalent` may hide; I think it it might be better to just remove these `validateEquivalent` things and rather go with `equals` and `hashCode` - there are pros and cons; but they also provide that things can be put inside a `Set` correctly...

so far the only pro is that they don't need to set the `outputColumn` properly...

File: processing/src/test/java/org/apache/druid/query/rowsandcols/semantic/FramedOnHeapAggregatableTest.java (hunk 5 above)
Comment (kgyrtkirk) [reply processing/src/test/java/org/apache/druid/query/rowsandcols/semantic/FramedOnHeapAggregatableTest.java:596]: I've added one - and it just passed because it does a `copy` in the [get()](https://github.com/apache/druid/blob/dff5bcb0a6c693e6fc8131e91c5029dbab132fa1/extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchBuildAggregator.java#L64)

**Review by imply-cheddar — APPROVED** [review]

**Comment by cheddar** [comment]
I went to squash-and-merge this PR, but all of the various commit messages are not really meaningful.  I looked to see if one of them would make a good commit message, but none of them seem to really work as a good message summarizing what the commit is doing.  Can you either include a proposed commit message that I can copy and paste into the Squash-and-Merge in the description of this PR?  And/or (and in the future) can you make sure to have at least one commit message that is intended to be used for the final upstream commit?
